File: src/nfc_writer/bridge.py
"""
文件名: bridge.py
创建日期: 2026-06-01
功能描述: pywebview JS 桥接层，将 NFC 服务方法包装为 JSON 安全的 API 供前端调用。
"""

from __future__ import annotations

from collections.abc import Callable
from typing import Any

from nfc_writer.nfc.service import NFCService


class NFCBridge:
    """暴露给 pywebview 前端的 JSON 安全 API 接口。"""

    def __init__(self, service: NFCService) -> None:
        """初始化桥接层。

        参数:
            service: NFC 服务实例。
        """
        self._service = service

    def ping(self) -> dict[str, Any]:
        """健康检查接口。"""
        return {"ok": True, "data": {"service": "nfc-writer"}}

    def status(self) -> dict[str, Any]:
        """获取当前 NFC 服务状态。"""
        return self._safe(self._service.status)

    def list_readers(self) -> dict[str, Any]:
        """列出所有可用的 NFC 读写器。"""
        return self._safe(self._service.list_readers)

    def connect(self, reader_id: str) -> dict[str, Any]:
        """连接到指定的 NFC 读写器。

        参数:
            reader_id: 目标读写器的标识符。
        """
        return self._safe(lambda: self._service.connect(reader_id))

    def disconnect(self) -> dict[str, Any]:
        """断开当前 NFC 读写器连接。"""
        return self._safe(self._service.disconnect)

    def read_tag(self) -> dict[str, Any]:
        """读取当前 NFC 标签数据。"""
        return self._safe(self._service.read_tag)

    def write_text(self, text: str) -> dict[str, Any]:
        """向当前 NFC 标签写入文本记录。

        参数:
            text: 要写入的文本内容。
        """
        return self._safe(lambda: self._service.write_text(text))

    @staticmethod
    def _safe(action: Callable[[], Any]) -> dict[str, Any]:
        """安全执行操作，捕获异常并统一返回格式。

        参数:
            action: 要执行的回调函数。

        返回:
            包含 ok 状态和 data 或 error 的字典。
        """
        try:
            return {"ok": True, "data": action()}
        except Exception as exc:
            return {"ok": False, "error": str(exc), "errorType": exc.__class__.__name__}

File: frontend/src/services/bridge.ts
import type { NFCAdapter, NFCReader, NFCStatus, NFCTag } from "../types/nfc";

export interface ApiResponse<T> {
  ok: boolean;
  data?: T;
  error?: string;
  errorType?: string;
}

export interface NFCBridgeApi {
  ping: () => Promise<ApiResponse<{ service: string }>>;
  status: () => Promise<ApiResponse<NFCStatus>>;
  list_adapters: () => Promise<ApiResponse<NFCAdapter[]>>;
  set_adapter: (adapterName: string) => Promise<ApiResponse<NFCStatus>>;
  list_readers: () => Promise<ApiResponse<NFCReader[]>>;
  connect: (readerId: string) => Promise<ApiResponse<NFCReader>>;
  disconnect: () => Promise<ApiResponse<NFCStatus>>;
  read_tag: () => Promise<ApiResponse<NFCTag>>;
  write_text: (text: string) => Promise<ApiResponse<NFCTag>>;
  write_uri: (uri: string) => Promise<ApiResponse<NFCTag>>;
}

const ADAPTERS: NFCAdapter[] = [
  { id: "pyscard", name: "标准 PC/SC", active: true },
  { id: "xinheng", name: "莘航 RFID", active: false },
];

// 浏览器预览模式下的本地状态
let currentAdapterId = "pyscard";

function getAdapterName(id: string): string {
  return ADAPTERS.find((a) => a.id === id)?.name ?? id;
}

const fallbackApi: NFCBridgeApi = {
  async ping() {
    return { ok: true, data: { service: "browser-preview" } };
  },
  async status() {
    return {
      ok: true,
      data: {
        adapter: currentAdapterId,
        adapterName: getAdapterName(currentAdapterId),
        connectedReaderId: null,
      },
    };
  },
  async list_adapters() {
    return {
      ok: true,
      data: ADAPTERS.map((a) => ({ ...a, active: a.id === currentAdapterId })),
    };
  },
  async set_adapter(adapterName: string) {
    const found = ADAPTERS.find((a) => a.id === adapterName);
    if (!found) {
      return { ok: false, error: `未知适配器: ${adapterName}` };
    }
    currentAdapterId = adapterName;
    return {
      ok: true,
      data: {
        adapter: currentAdapterId,
        adapterName: getAdapterName(currentAdapterId),
        connectedReaderId: null,
      },
    };
  },
  async list_readers() {
    return { ok: true, data: [] };
  },
  async connect(readerId: string) {
    return { ok: true, data: { id: readerId, name: readerId, connected: true } };
  },
  async disconnect() {
    return {
      ok: true,
      data: {
        adapter: currentAdapterId,
        adapterName: getAdapterName(currentAdapterId),
        connectedReaderId: null,
      },
    };
  },
  async read_tag() {
    return { ok: true, data: { uid: "", records: [] } };
  },
  async write_text(text: string) {
    return { ok: true, data: { uid: "", records: [text] } };
  },
  async write_uri(uri: string) {
    return { ok: true, data: { uid: "", records: [uri] } };
  },
};

export function getBridge(): NFCBridgeApi {
  return window.pywebview?.api ?? fallbackApi;
}

export function unwrap<T>(response: ApiResponse<T>): T {
  if (!response.ok || response.data === undefined) {
    throw new Error(response.error ?? "Bridge call failed.");
  }
  return response.data;
}
